compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -455,24 +665,81 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
    */
   def replace(param: TypeParamRef, tp: Type)(using Context): OrderingConstraint =
     val replacement = tp.dealiasKeepAnnots.stripTypeVar
-    if param == replacement then this.checkNonCyclic()
+    if param == replacement then this.checkWellFormed()
     else
       assert(replacement.isValueTypeOrLambda)
-      var current =
-        if isRemovable(param.binder) then remove(param.binder)
-        else updateEntry(this, param, replacement)
-
-      def removeParam(ps: List[TypeParamRef]) = ps.filterConserve(param ne _)
 
-      def replaceParam(tp: Type, atPoly: TypeLambda, atIdx: Int): Type =
-        current.ensureNonCyclic(atPoly.paramRefs(atIdx), tp.substParam(param, replacement))
-
-      current.foreachParam { (p, i) =>
-        current = boundsLens.map(this, current, p, i, replaceParam(_, p, i))
-        current = lowerLens.map(this, current, p, i, removeParam)
-        current = upperLens.map(this, current, p, i, removeParam)
-      }
-      current.checkNonCyclic()
+      val replacedTypeVar = typeVarOfParam(param)
+      //println(i"replace $param with $replacement in $this")
+
+      def mapReplacedTypeVarTo(to: Type) = new TypeMap:
+        override def apply(t: Type): Type =
+          if (t eq replacedTypeVar) && t.exists then to else mapOver(t)
+
+      var current = this
+
+      def removeParamFrom(ps: List[TypeParamRef]) =
+        ps.filterConserve(param ne _)
+
+      for lo <- lower(param) do
+        current = upperLens.map(this, current, lo, removeParamFrom)
+      for hi <- upper(param) do
+        current = lowerLens.map(this, current, hi, removeParamFrom)
+
+      def replaceParamIn(other: TypeParamRef) =
+        val oldEntry = current.entry(other)
+        val newEntry = current.ensureNonCyclic(other, oldEntry.substParam(param, replacement))
+        current = boundsLens.update(this, current, other, newEntry)
+        var oldDepEntry = oldEntry
+        var newDepEntry = newEntry
+        replacedTypeVar match
+          case tvar: TypeVar =>
+            if tvar.isInstantiated
```

Comment:
```suggestion
            if tvar.inst.exists
```
(I believe we don't need to check `instType(tvar)` since at this point we haven't called updateEntry to instantiate the param in the constraint yet)